### PR TITLE
Implement CDS on FHIR in CDS Hooks

### LINKED ISSUE #5238: Implement CDS on FHIR in CDS Hooks
Implement the ability for CDS Services to be generated and registered from PlanDefinitions in the DAO that can be executed with the $apply operation and return CDS Cards in accordance with Clinical Reasoning CDS on FHIR: https://build.fhir.org/clinicalreasoning-cds-on-fhir.html

### PR DESCRIPTION
**Pull request by barhodes**
Closes #5238

### CODE REVIEW

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceDstu3.java [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+			return ret;
+		}
+	}
+
+	public PrefetchUrlList getPrefetchUrlList(PlanDefinition thePlanDefinition) {
+		PrefetchUrlList prefetchList = new PrefetchUrlList();
+		if (thePlanDefinition == null) return null;
+		if (!isEca(thePlanDefinition)) return null;
+		Library library = resolvePrimaryLibrary(thePlanDefinition);
+		// TODO: resolve data requirements
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceDstu3.java:201]: Do you plan on addressing this TODO in a later PR?

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceR4.java [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+			return ret;
+		}
+	}
+
+	public PrefetchUrlList getPrefetchUrlList(PlanDefinition thePlanDefinition) {
+		PrefetchUrlList prefetchList = new PrefetchUrlList();
+		if (thePlanDefinition == null) return null;
+		if (!isEca(thePlanDefinition)) return null;
+		Library library = resolvePrimaryLibrary(thePlanDefinition);
+		// TODO: resolve data requirements
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceR4.java:188]: Do you plan on addressing this TODO in a future PR?

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceR5.java [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+			return ret;
+		}
+	}
+
+	public PrefetchUrlList getPrefetchUrlList(PlanDefinition thePlanDefinition) {
+		PrefetchUrlList prefetchList = new PrefetchUrlList();
+		if (thePlanDefinition == null) return null;
+		if (!isEca(thePlanDefinition)) return null;
+		Library library = resolvePrimaryLibrary(thePlanDefinition);
+		// TODO: resolve data requirements
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceR5.java:190]: Same question about TODO

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceDstu3.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+	public FhirVersionEnum getFhirVersion() {
+		return FhirVersionEnum.DSTU3;
+	}
+
+	public Repository getRepository() {
+		return myRepository;
+	}
+
+	public Parameters encodeParams(CdsServiceRequestJson theJson) {
+		// CanonicalType canonical = ;
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceDstu3.java:80]: Do you plan to leave these code snippets commented out?

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceDstu3.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+		}
+		resourceMap.forEach((key, value) -> prefetchResources.addEntry().setResource(value));
+		return prefetchResources;
+	}
+
+	public CdsServiceResponseJson encodeResponse(Object theResponse) {
+		assert theResponse instanceof CarePlan;
+		myResponse = (CarePlan) theResponse;
+		CdsServiceResponseJson serviceResponse = new CdsServiceResponseJson();
+		Reference requestGroupRef = myResponse.getActivity().get(0).getReference();
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceDstu3.java:179]: What if myResponse.getActivity() and getDefinition() are empty?

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR4.java [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+		return FhirVersionEnum.R4;
+	}
+
+	public Repository getRepository() {
+		return myRepository;
+	}
+
+	public Parameters encodeParams(CdsServiceRequestJson theJson) {
+		Parameters parameters = parameters()
+				// .addParameter(part(APPLY_PARAMETER_CANONICAL, canonical))
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR4.java:82]: Do you plan on leaving this commented out?

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR5.java [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+	}
+
+	public Repository getRepository() {
+		return myRepository;
+	}
+
+	public Parameters encodeParams(CdsServiceRequestJson theJson) {
+		// CanonicalType canonical = ;
+		Parameters parameters = parameters()
+				// .addParameter(part(APPLY_PARAMETER_CANONICAL, canonical))
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR5.java:83]: Similar question about commented out code

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR4.java [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+		resourceMap.forEach((key, value) -> prefetchResources.addEntry().setResource(value));
+		return prefetchResources;
+	}
+
+	public CdsServiceResponseJson encodeResponse(Object theResponse) {
+		assert theResponse instanceof Bundle;
+		myResponseBundle = (Bundle) theResponse;
+		myServiceResponse = new CdsServiceResponseJson();
+		RequestGroup mainRequest =
+				(RequestGroup) myResponseBundle.getEntry().get(0).getResource();
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR4.java:181]: What happens if getEntry() and getInstantiatesCanonical() are empty?

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR4.java [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+		String[] split = reference.split("/");
+		String id = reference.contains("/") ? split[1] : reference;
+		String resourceType = reference.contains("/") ? split[0] : theReference.getType();
+		return myResponseBundle.getEntry().stream()
+				.filter(entry -> entry.hasResource()
+						&& entry.getResource().getResourceType().toString().equals(resourceType)
+						&& entry.getResource().getIdPart().equals(id))
+				.map(entry -> entry.getResource())
+				.collect(Collectors.toList())
+				.get(0);
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR4.java:331]: What happens if this List is empty?

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR5.java [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+
+	protected IBaseResource resolveResource(Reference theReference) {
+		String reference = theReference.getReference();
+		String id = reference.contains("/") ? reference.split("/")[1] : reference;
+		return myResponseBundle.getEntry().stream()
+				.filter(entry ->
+						entry.hasResource() && entry.getResource().getId().equals(id))
+				.map(entry -> entry.getResource())
+				.collect(Collectors.toList())
+				.get(0);
```
Comment (lukedegruchy) [inline hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR5.java:332]: Same question about empty

**Review by lukedegruchy — APPROVED** [review]
Approved but with some NPE/empty List checking concerns

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR4.java (hunk 9 above)
Comment (barhodes) [reply hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/CdsCrServiceR4.java:331]: Shouldn't ever happen.  It means the $apply operation is adding references to resources without adding those resources to the bundle.  Changed the logic to handle the case anyway.

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryElementDstu3.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+	protected PlanDefinition myPlanDefinition;
+	protected PrefetchUrlList myPrefetchUrlList;
+
+	public CrDiscoveryElementDstu3(PlanDefinition thePlanDefinition, PrefetchUrlList thePrefetchUrlList) {
+		myPlanDefinition = thePlanDefinition;
+		myPrefetchUrlList = thePrefetchUrlList;
+	}
+
+	public CdsServiceJson getCdsServiceJson() {
+		if (myPlanDefinition == null
```
[the remark this answers is not in the document]
Comment (barhodes) [reply hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryElementDstu3.java:39]: Yes, thank you for catching this!

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceR4.java (hunk 2 above)
Comment (barhodes) [reply hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceR4.java:188]: Yes, on the backlog

File: hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceR5.java (hunk 3 above)
Comment (barhodes) [reply hapi-fhir-server-cds-hooks/src/main/java/ca/uhn/hapi/fhir/cdshooks/svc/cr/discovery/CrDiscoveryServiceR5.java:190]: Yes, on the backlog

**Pull request merged**